lib/rpmts.c

```diff
@@ -575,29 +575,79 @@ rpmRC rpmtsImportHeader(rpmtxn txn, Header h, rpmFlags flags)
     return rc;
 }
 
-static rpmRC rpmtsImportFSKey(rpmtxn txn, Header h, rpmFlags flags)
+static rpmRC rpmtsImportFSKey(rpmtxn txn, Header h, rpmFlags flags, int replace)
 {
     rpmRC rc = RPMRC_FAIL;
     char *keyfmt = headerFormat(h, "%{nvr}.key", NULL);
     char *keyval = headerGetAsString(h, RPMTAG_DESCRIPTION);
     char *path = rpmGenPath(rpmtsRootDir(txn->ts), "%{_keyringpath}/", keyfmt);
+    char *tmppath = NULL;
 
-    FD_t fd = Fopen(path, "wx");
+    if (replace) {
+	rasprintf(&tmppath, "%s.new", path);
+	unlink(tmppath);
+    }
+
+    FD_t fd = Fopen(tmppath ? tmppath : path, "wx");
     if (fd) {
 	size_t keylen = strlen(keyval);
 	if (Fwrite(keyval, 1, keylen, fd) == keylen)
 	    rc = RPMRC_OK;
 	Fclose(fd);
     }
+    if (!rc && tmppath && rename(tmppath, path) != 0)
+	rc = RPMRC_FAIL;
 
     if (rc) {
 	rpmlog(RPMLOG_ERR, _("failed to import key: %s: %s\n"),
-		path, strerror(errno));
+		tmppath ? tmppath : path, strerror(errno));
+	if (tmppath)
+	    unlink(tmppath);
+    }
+    
+    if (!rc && replace) {
+	/* find and delete the old pubkey entry */
```

Comment:
What's the use case for these semantics? (Deleting the old version of a certificate and replacing it with a new version.)

Reply:
see above.